### PR TITLE
Add SQLExecuteQueryOperator

### LINKED ISSUE #25259: Add `SqlExecuteQueryOperator`
### Body

There has been previous attempt to add this operator in https://github.com/apache/airflow/pull/10011
But now after https://github.com/apache/airflow/pull/23971 is completed this should be much easier as `DbApiHook` has unified run method that works for all providers (with some configuration of parameters like `split_statements` etc..)

We can now explore again the option of `SqlExecuteQueryOperator`
There are two approaches that I can think of:
1. Creating a `BaseSqlExecuteQueryOperator` that all Sql related operators (`PostgresOperator`, `MySqlOperator`, `SnowflakeOperator`) will inherit from.
2. Deprecate the current individual operators in favor of `SqlExecuteQueryOperator` (just like https://github.com/apache/airflow/pull/10011 attempted to do)

Personally I prefer to try the 2nd approach first (It should be similar to what we do in `SqlSensor` or `SqlBranchOperator` the issue to resolve is how to handle the batch queries in some of the DBs and that is now should be possible thanks to https://github.com/apache/airflow/pull/23971  in `SqlSensor` / `SqlBranchOperator` executing batch queries was never an issue as they are expected to execute only 1 SQL statement )

FYI @potiuk @kazanzhy 

### Committer

- [X] I acknowledge that I am a maintainer/committer of the Apache Airflow project.

### PR DESCRIPTION
**Pull request by kazanzhy**
There are many changes here that potentially can break backward compatibility, therefore should be tested.
Work in progress and this PR is created for discussions if some things could be implemented better.

closes: #25259

### CODE REVIEW

File: airflow/providers/common/sql/operators/sql.py [hunk 1]
```diff
@@ -112,6 +124,61 @@ def get_db_hook(self) -> DbApiHook:
         return self._hook
 
 
+class SQLExecuteQueryOperator(BaseSQLOperator):
```
Comment (vincbeck) [inline airflow/providers/common/sql/operators/sql.py:168]: Dont forget to add documentation for this new operator.

**Review by vincbeck — CHANGES_REQUESTED** [review]

File: airflow/providers/apache/drill/operators/drill.py [hunk 2]
```diff
@@ -15,16 +15,16 @@
...
+import warnings
+from typing import TYPE_CHECKING, Sequence
 
-from airflow.models import BaseOperator
-from airflow.providers.apache.drill.hooks.drill import DrillHook
+from airflow.providers.common.sql.operators.sql import SQLExecuteQueryOperator
 
 if TYPE_CHECKING:
-    from airflow.utils.context import Context
+    pass
```
Comment (vincbeck) [inline airflow/providers/apache/drill/operators/drill.py]: This if looks useless to me

File: airflow/providers/common/sql/operators/sql.py [hunk 3]
```diff
@@ -63,13 +75,13 @@ class BaseSQLOperator(BaseOperator):
...
     retrieve the DB hook based on connection type.
-    You can custom the behavior by overriding the .get_db_hook() method.
+    You can customize the behavior by overriding the .get_db_hook() method.
     """
 
     def __init__(
         self,
         *,
-        conn_id: Optional[str] = None,
+        conn_id: str,
```
Comment (vincbeck) [inline airflow/providers/common/sql/operators/sql.py]: This seems not backward compatible to me

File: airflow/providers/common/sql/operators/sql.py [hunk 4]
```diff
@@ -112,6 +124,61 @@ def get_db_hook(self) -> DbApiHook:
...
 
 
+class SQLExecuteQueryOperator(BaseSQLOperator):
+    """
+    Executes SQL code in a specific database
+    :param sql: the SQL code or string pointing to a template file to be executed (templated).
+    File must have a '.sql' extensions.
+    :param handler: (optional) the function that will be applied to the cursor.
+    :param parameters: (optional) the parameters to render the SQL query with.
+    :param conn_id: reference to a specific database
```
Comment (vincbeck) [inline airflow/providers/common/sql/operators/sql.py]: Some parameters are missing for the docstring

File: airflow/providers/common/sql/operators/sql.py [hunk 5]
```diff
@@ -112,6 +124,61 @@ def get_db_hook(self) -> DbApiHook:
...
+                return_last=self.return_last,
+            )
+        else:
+            output = hook.run(self.sql, autocommit=self.autocommit, parameters=self.parameters)
+
+        if hasattr(self, '_process_output'):
+            for out in output:
+                self._process_output(*out)
+
+        return output
```
Comment (vincbeck) [inline airflow/providers/common/sql/operators/sql.py]: This if is not really clear to me. Why not splitting the statements if `do_xcom_push` is `False`? How is that related? Same for `return_last`. I would do it differently:

```suggestion
            output = hook.run(
                self.sql,
                autocommit=self.autocommit,
                parameters=self.parameters,
                handler=self.handler,
                split_statements=self.split_statements,
                return_last=self.return_last,
            )

        if hasattr(self, '_process_output'):
            for out in output:
                self._process_output(*out)

        return output
```

Comment (vincbeck) [reply airflow/providers/common/sql/operators/sql.py]: After, it is up to the user to use the returned value or not (using Xcom)

File: airflow/providers/exasol/operators/exasol.py [hunk 6]
```diff
@@ -15,16 +15,16 @@
...
+import warnings
+from typing import TYPE_CHECKING, Sequence, Optional
 
-from airflow.models import BaseOperator
-from airflow.providers.exasol.hooks.exasol import ExasolHook
+from airflow.providers.common.sql.operators.sql import SQLExecuteQueryOperator
 
 if TYPE_CHECKING:
-    from airflow.utils.context import Context
+    pass
```
Comment (vincbeck) [inline airflow/providers/exasol/operators/exasol.py]: Is it needed?

File: airflow/providers/databricks/operators/databricks_sql.py [hunk 7]
```diff
@@ -80,53 +80,44 @@ class DatabricksSqlOperator(BaseOperator):
...
+            'http_headers': http_headers,
+            'catalog': catalog,
+            'schema': schema,
+            'caller': "DatabricksSqlOperator",
+            **client_parameters,
+            **hook_params,
+        }
+
+    def get_db_hook(self) -> DatabricksSqlHook:
+        return DatabricksSqlHook(self.databricks_conn_id, **self.hook_params)
```
Comment (vincbeck) [inline airflow/providers/databricks/operators/databricks_sql.py:119]: In the other operators, defining only the `hook_params` seem enough. e.g. `kwargs['hook_params'] = {'schema': schema, **hook_params}`. Why in this case you need to override `get_db_hook`?

File: airflow/providers/jdbc/operators/jdbc.py [hunk 8]
```diff
@@ -15,18 +15,16 @@
...
+from typing import TYPE_CHECKING, Sequence
 
-from typing import TYPE_CHECKING, Any, Callable, Iterable, Mapping, Optional, Sequence, Union
-
-from airflow.models import BaseOperator
-from airflow.providers.common.sql.hooks.sql import fetch_all_handler
-from airflow.providers.jdbc.hooks.jdbc import JdbcHook
+from airflow.providers.common.sql.operators.sql import SQLExecuteQueryOperator
 
 if TYPE_CHECKING:
```
Comment (vincbeck) [inline airflow/providers/jdbc/operators/jdbc.py]: Needed?

File: airflow/providers/postgres/operators/postgres.py [hunk 9]
```diff
@@ -15,19 +15,19 @@
...
+from typing import TYPE_CHECKING, Mapping, Optional, Sequence
 
 from psycopg2.sql import SQL, Identifier
 
-from airflow.models import BaseOperator
-from airflow.providers.postgres.hooks.postgres import PostgresHook
+from airflow.providers.common.sql.operators.sql import SQLExecuteQueryOperator
 from airflow.www import utils as wwwutils
 
 if TYPE_CHECKING:
```
Comment (vincbeck) [inline airflow/providers/postgres/operators/postgres.py]: Needed?

File: airflow/providers/trino/operators/trino.py [hunk 10]
```diff
@@ -16,19 +16,19 @@
...
+import warnings
+from typing import TYPE_CHECKING, Any, Sequence
 
 from trino.exceptions import TrinoQueryError
 
-from airflow.models import BaseOperator
+from airflow.providers.common.sql.operators.sql import SQLExecuteQueryOperator
 from airflow.providers.trino.hooks.trino import TrinoHook
 
 if TYPE_CHECKING:
```
Comment (vincbeck) [inline airflow/providers/trino/operators/trino.py]: Needed?

File: airflow/providers/vertica/operators/vertica.py [hunk 11]
```diff
@@ -15,16 +15,16 @@
...
 # under the License.
-from typing import TYPE_CHECKING, Any, Iterable, Sequence, Union
+import warnings
+from typing import TYPE_CHECKING, Any, Sequence
 
-from airflow.models import BaseOperator
-from airflow.providers.vertica.hooks.vertica import VerticaHook
+from airflow.providers.common.sql.operators.sql import SQLExecuteQueryOperator
 
 if TYPE_CHECKING:
```
Comment (vincbeck) [inline airflow/providers/vertica/operators/vertica.py]: Needed?

File: airflow/providers/common/sql/operators/sql.py (hunk 1 above)
Comment (vincbeck) [inline airflow/providers/common/sql/operators/sql.py:168]: Please create unit tests for this operator. You should not rely on unit tests made for all the different DB operators because at some point we are going to remove them (since they are now deprecated). Along the operators we'll remove, we'll remove as well the unit tests associated to them

**Review by vincbeck — CHANGES_REQUESTED** [review]

**Comment by kazanzhy** [comment]
@vincbeck thanks for the suggestions.
Currently, I'm trying to fix tests, because before that each hook was patched by mock and now there're a lot of issues with that

File: airflow/providers/common/sql/operators/sql.py (hunk 5 above)
Comment (kazanzhy) [reply airflow/providers/common/sql/operators/sql.py]: Here is one interesting thing.
If we set `None` as the default handler then results won't be returned. 
Even for `do_xcom_push=True`.
In another case, if we want to keep `fetch_all_handler` as the default handler then we can't pass it for non-SELECT statements because it raises an error.

File: airflow/providers/databricks/operators/databricks_sql.py (hunk 7 above)
Comment (kazanzhy) [reply airflow/providers/databricks/operators/databricks_sql.py:119]: By default, Connection will return `airflow.providers.databricks.hooks.databricks.DatabricksHook`. 
And I didn't found the solution how to fix it

**Review by potiuk — COMMENTED** [review]
Looks coo, but you should increase common.sql to 1.3 and add 1.3 as a dependency on all providers affected @kazanzhy

**Comment by potiuk** [comment]
(and rebase the change after __future__.annotations change too :)

**Comment by vincbeck** [comment]
Where are you at @kazanzhy. Feel free to share the status of this PR, we might also be able to help you in case you need it so feel free to reach out :) It looks like your local modification reverted future.annotations made by Jarek. I encourage you to run the static-check locally with the command `breeze static-checks`

**Comment by kazanzhy** [comment]
@vincbeck
Sorry about that. It was many changes after my commit so I'm fixing rebase issues

File: airflow/providers/common/sql/operators/sql.py [hunk 12]
```diff
@@ -162,6 +164,68 @@ def get_db_hook(self) -> DbApiHook:
...
+
+    def __init__(
+        self,
+        *,
+        sql: str | list[str],
+        autocommit: bool = False,
+        parameters: Mapping | Iterable | None = None,
+        handler: Callable[[Any], Any] = fetch_all_handler,
+        split_statements: bool = False,
+        return_last: bool = True,
```
Comment (eladkal) [inline airflow/providers/common/sql/operators/sql.py:197]: This parameter works only with `do_xcom_push=True`.
I have two thoughts on this:
1. If `return_last=True` should we under the hood set  `do_xcom_push=True` ?
2. Assuming we go with (1) then probably the default of `return_last` should be False?

My goal here is to prevent the undesired state of  `do_xcom_push=False`  with `return_last=True`.
If user set `return_last=True` we know for sure xcom needs to be pushed.
if user set `return_last=False` we don't know what is the user wish for xcom so we need to also look for `do_xcom_push` value.

Comment (kazanzhy) [reply airflow/providers/common/sql/operators/sql.py:197]: Based on my previous PR which had broke compatibility I decided to avoid it here.

Talking about the `return_last` it works only when `do_xcom_push=True` (by default True) and `split_statements=True` (by default False).
Probably it should be renamed to `return_last_if_split_statements` :)
Why it's `True` by default?
I think when we added `split_statements` we wanted to keep the same returning type independent what were passed there.
https://github.com/apache/airflow/pull/23971#discussion_r922847638

If the user has in the legacy code, let's say, PostgresOperator. By default 
it has `do_xcom_push = True`, `split_statements = False` and `return_last = True`. 
https://github.com/apache/airflow/blob/69da98cdb194de3544368b6bd7c47dcc7ace8814/airflow/models/baseoperator.py#L735
https://github.com/apache/airflow/blob/69da98cdb194de3544368b6bd7c47dcc7ace8814/airflow/providers/common/sql/hooks/sql.py#L261

So if the's a string as an SQL parameter. then result of this query will be returned. If the list of strings is passed as an SQL parameter, then wil be returned list of results of each statement. Of course in case if handler were passed. I think here we discussed it here https://github.com/apache/airflow/pull/23971#discussion_r885115775

**Review by eladkal — APPROVED** [review]
LGTM

File: airflow/providers/common/sql/operators/sql.py [hunk 13]
```diff
@@ -162,6 +164,68 @@ def get_db_hook(self) -> DbApiHook:
...
+    :param autocommit: (optional) if True, each command is automatically committed (default: False).
+    :param parameters: (optional) the parameters to render the SQL query with.
+    :param handler: (optional) the function that will be applied to the cursor.
+    :param split_statements: (optional) if split single SQL string into statements (default: False).
+    :param return_last: (optional) if return the result of only last statement (default: True).
+    """
+
+    template_fields: Sequence[str] = ('sql', 'parameters')
+    template_ext: Sequence[str] = ('.sql',)
+    ui_color = '#cdaaed'
```
Comment (eladkal) [inline airflow/providers/common/sql/operators/sql.py]: Lets add
`template_fields_renderers = {"sql": "sql"}`

**Comment by vincbeck** [comment]
Documentation is still missing :)

**Comment by kazanzhy** [comment]
Thanks @vincbeck, added docs. 
This operator is pretty simple bu if you have something more to add feel free for suggestions.

**Review by vincbeck — APPROVED** [review]
LGTM

**Comment by eladkal** [comment]
@kazanzhy there are some test failures

**Review by eladkal — APPROVED** [review]
Last call for further review, if no more comments I'll merge this PR tomorrow
great job @kazanzhy !

**Comment by fritz-astronomer** [comment]
I like the change - however completely removing the `execute` feels really aggressive. Shouldn't we deprecate and then remove later? This will break a *lot* of workflows if anyone's dependency get's bumped without them knowing about the change.

**Comment by eladkal** [comment]
> I like the change - however completely removing the `execute` feels really aggressive. Shouldn't we deprecate notice and then remove later? This will break a _lot_ of workflows if anyone's dependency get's bumped without them knowing about the change.

This PR is backward compatible.
The operators are now inheriting from `SQLExecuteQueryOperator` so they use this operator `execute()`

**Comment by fritz-astronomer** [comment]
Ah that makes a lot more sense! Apologies! Thank you for the clarification, I see that now.

**Comment by kazanzhy** [comment]
@fritz-astronomer I'll be so grateful if you will help me test these changes when RC will be ready

**Comment by syedahsn** [comment]
This PR is causing the `RedshiftSQLOperator` to fail with the message
```
AttributeError: 'RedshiftSQLOperator' object has no attribute 'redshift_conn_id'
--
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
File "/opt/airflow/airflow/models/taskinstance.py", line 1450, in _run_raw_task
self._execute_task_with_callbacks(context, test_mode)
File "/opt/airflow/airflow/models/taskinstance.py", line 1569, in _execute_task_with_callbacks
task_orig = self.render_templates(context=context)
File "/opt/airflow/airflow/models/taskinstance.py", line 2184, in render_templates
original_task.render_template_fields(context)
File "/opt/airflow/airflow/models/baseoperator.py", line 1196, in render_template_fields
self._do_render_template_fields(self, self.template_fields, context, jinja_env, set())
File "/opt/airflow/airflow/utils/session.py", line 75, in wrapper
return func(*args, session=session, **kwargs)
File "/opt/airflow/airflow/models/abstractoperator.py", line 405, in _do_render_template_fields
f"{attr_name!r} is configured as a template field "
AttributeError: 'redshift_conn_id' is configured as a template field but RedshiftSQLOperator does not have this attribute.
```
The reason seems to be the fact that the `redshift_conn_id` attribute in the `RedshiftSQLOperator` was a tempate field, but now that it no longer exists, it is causing issues.
In terms of maintaining backward compatibility, should we introduce the attribute back to the `RedshiftSQLOperator` to resolve the issue? Or is there a better approach? I'd love to hear other opinions on this. Thanks!
@eladkal @kazanzhy

**Comment by eladkal** [comment]
We can modify the deprecated class so it will be backward compatible and add nodify the new operator to template conn_id
Would you like to raise PR?

**Comment by syedahsn** [comment]
Sure, I'll get started on that.
